File: backend/internal/models/attendance.go
package models

type CreateAttendanceSessionRequest struct {
	ClassID string `json:"class_id" binding:"required"`
	Date    string `json:"date" binding:"required"`
}

type AttendanceRecord struct {
	StudentID string `json:"student_id" binding:"required"`
	Status    string `json:"status" binding:"required"`
	Note      string `json:"note"`
}

type MarkAttendanceRequest struct {
	Records []AttendanceRecord `json:"records" binding:"required"`
}

type AttendanceSessionResponse struct {
	ID        string `json:"id"`
	ClassID   string `json:"class_id"`
	TakenBy   string `json:"taken_by"`
	Date      string `json:"date"`
	CreatedAt string `json:"created_at"`
}

type AttendanceRecordResponse struct {
	ID           string `json:"id"`
	SessionID    string `json:"session_id"`
	StudentID    string `json:"student_id"`
	StudentName  string `json:"student_name"`
	StudentIndex string `json:"student_index"`
	Status       string `json:"status"`
	Note         string `json:"note"`
}

type AttendanceSummaryResponse struct {
	TotalDays int64 `json:"total_days"`
	Present   int64 `json:"present"`
	Absent    int64 `json:"absent"`
	Late      int64 `json:"late"`
	Excused   int64 `json:"excused"`
}

File: frontend/src/services/attendance.ts
import api from "../lib/api";


export interface AttendanceSession {
  id: string;
  class_id: string;
  taken_by: string;
  date: string;
  created_at: string | null;
}

export interface CreateSessionRequest {
  class_id: string;
  date: string; // YYYY-MM-DD
}

export interface AttendanceRecordRow {
  id: string;
  session_id: string;
  student_id: string;
  student_name: string;
  student_index: string;
  status: "present" | "absent" | "late" | "excused";
  note: string | null;
}

export interface MarkAttendanceRequest {
  records: {
    student_id: string;
    status: "present" | "absent" | "late" | "excused";
    note?: string;
  }[];
}

export interface DailySession extends AttendanceSession {
  class_name: string;
  grade_name: string;
  teacher_name: string;
  enrolled_count: number;
  marked_count: number;
}

export const attendanceApi = {
  listSessionsByClass: (classId: string) =>
    api
      .get<AttendanceSession[]>(`/classes/${classId}/attendance/sessions`)
      .then((r) => r.data),

  listSessionsByDate: (date: string) =>
    api
      .get<DailySession[]>("/attendance/sessions", { params: { date } })
      .then((r) => r.data),

  createSession: (data: CreateSessionRequest) =>
    api.post<AttendanceSession>("/attendance/sessions", data).then((r) => r.data),

  getSession: (id: string) =>
    api.get<AttendanceSession>(`/attendance/sessions/${id}`).then((r) => r.data),

  listRecords: (sessionId: string) =>
    api
      .get<AttendanceRecordRow[]>(`/attendance/sessions/${sessionId}/records`)
      .then((r) => r.data),

  markAttendance: (sessionId: string, data: MarkAttendanceRequest) =>
    api
      .post(`/attendance/sessions/${sessionId}/records`, data)
      .then((r) => r.data),

  // cascades to every attendance record already written for the session
  deleteSession: (id: string) =>
    api.delete(`/attendance/sessions/${id}`).then((r) => r.data),
};
